### ISSUE TITLE
Hundreds of thousands files in /root/.config/gops

### ISSUE THREAD

**Issue by dkerwin** [cilium/cilium#11455]
## Bug report

**General Information**

- Cilium version (run `cilium version`)

```
Client: 1.7.1 1ba69e67b 2020-03-04T15:44:41-08:00 go version go1.13.8 linux/amd64
Daemon: 1.7.1 1ba69e67b 2020-03-04T15:44:41-08:00 go version go1.13.8 linux/amd64
```

- Kernel version (run `uname -a`)

```
Linux mt16 5.4.0-0.bpo.3-amd64 #1 SMP Debian 5.4.13-1~bpo10+1 (2020-02-07) x86_64 x86_64 x86_64 GNU/Linux
```

- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...)

```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T23:35:15Z", GoVersion:"go1.14.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-11T18:07:13Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
```

**How to reproduce the issue**

1. Let cilium run
2. Observe files being created in /root/.config/gops

In 24h around 18000 files were created. This issue came to my attention when the filesystem was slowly running out of nodes. Example content of one of the files:

```
root@mt16:~/.config/gops# cat 1813287
40037
```

**Comment by tklauser** [comment]
Thanks for the report.

These files are created by the gops agent (which cilium integrates for runtime diagnostics). It looks like it isn't properly cleaning them up in case the process is killed with a signal other than `SIGINT`. I'll provide a fix to the gops package and make sure it is vendored back into the cilium repo which should hopefully fix this behavior.

**Comment by dkerwin** [comment]
Thanks for the explanation. What if the agent just continues to run? In the long run it would create the same issue. Is the gops agent not running constantly?

**Comment by tklauser** [comment]
As long as the agent just continues to run we shouldn't hit this issue as the file is only created once on startup. The gops agent is included as a package within cilium, so it is part of the cilium binary. What puzzles me a bit about the reported case is why there are so many files within a rather short amount of time. Is there anything in the logs indicating that the cilium agent is respawned? Or are you restarting it on purpose?

As an aside: Upon closer inspection, it also looks like we're not setting the proper option when starting the gops agent, i.e. we should be invoking it with `agent.Options{ShutdownCleanup: true}`. I'll include that part in my fix as well.

**Comment by dkerwin** [comment]
I see no agent restarts and the logs don't show something suspicious either. The files are actually created non stop:

```
root@mt16:~/.config/gops# ls -l | wc -l && sleep 30 && ls -l | wc -l
20272
20278
```

Do it doesn't look like a shutdown-only issue to me.

**Comment by tklauser** [comment]
Thanks for checking. Will investigate further on what might be causing this.

**Comment by tklauser** [comment]
@dkerwin I think PR #11471 should solve the issue you're seeing as it makes sure to properly tear down the gops socket and delete the files in `/root/.config/gops` upon exit.

Though, I still cannot explain entirely why you see so many files within such a short time span. One hypothesis would be that this is because of the gops agent in `cilium-cni` which might get invoked a lot if lots of pods are added/removed.

Do you have any way of checking the fix from PR #11471 in your setup and verify that it solves the problem? Otherwise you'll probably have to wait for the next backport release which includes this fix.

**Comment by tklauser** [comment]
We think that this is fixed by #11471 which was backported by #11524 to the `v1.7` branch. It is now included in the recent `v1.7.4` release. Thus, I'm closing this.

@dkerwin Please feel free to reopen this issue if you still observe the reported behavior. Thanks.
